### PR TITLE
build: add support for LTO and PGO to the building system

### PR DESCRIPTION
**Pull request by tchaikov**
This changeset ports LTO and PGO support from scylla-enterprise.git to scylladb.git. 

Add support for Link-Time Optimization (LTO) and Profile-Guided Optimization (PGO) 
to improve performance. LTO provides ~7% performance gain and enables crucial 
binary layout optimizations for PGO.

LTO Changes:
- Add `-flto` flag to compile and link steps
- Use `-ffat-lto-objects` to generate both LLVM IR and machine code
- Enable cross-object optimization while maintaining fast test linking

PGO Implementation:
- Implement three-stage build process:
  1. Context-free profiling (`-fprofile-generate`)
  2. Context-sensitive profiling (`-fprofile-use` + `-fcs-profile-generate`)
  3. Final optimization using merged profiles
- Add release-pgo and release-cs-pgo build stages
- Integrate with ninja build system
- Stages can be enabled independently

Profile Management:
- Add `pgo/pgo.py` for workload profile collection
- Store default profile in `pgo/profiles/profile.profdata.xz` using Git LFS
- Add configure.py integration for profile detection and validation
- Support custom profiles via `--use-profile` flag
- Add profile regeneration script

Both optimizations are recommended for maximum performance, though each PGO 
stage adds a full build cycle. Future optimization may allow dropping one 
PGO stage if performance impact is minimal.

---

this is a forward port, hence no need to backport.

### CODE REVIEW

**Comment by tchaikov** [comment]
v2:

- restructure the commits to squash the fixes into the commit fixed by it.
- add quotes around expressions like `"$(dirname $(realpath $0))"` to prevent word splitting in bash scripts
- add `&` before `2` in
    ```sh
    echo "Error: $SCRIPT_PATH should be ran with $PROJECT_BASE instead of $WORKING_DIR as the working directory" >2
    ```

**Comment by avikivity** [comment]
Please adjust the cover letter to be a normal feature introduction, and check again with diff that it is complete.

**Comment by tchaikov** [comment]
v3:

- updated the cover letter to be a normal feature introduction
- compared `configure.py`, related CMake scripts and `pgo` subdirectory with enterprise branch, added the missing changes:

  * added `profile_deps` dependency to `build {obj}` rule.
  * corrected the "description" of `xz_uncompress` rule

**Comment by tchaikov** [comment]
> Please adjust the cover letter to be a normal feature introduction, and check again with diff that it is complete.

done and done.

**Comment by tchaikov** [comment]
@avikivity Hi Avi, when comparing `configure.py`, i noticed that you changed the "optimization-level" of "debug" build from "g" to "0" in https://github.com/scylladb/scylla-enterprise/commit/d0a60b1d6ffc5f73441fa2723ef89e478be18e2e#diff-4d5f3192809ec1b9add6b33007e0c50031ad9a0a2f3f55a481b506468824db2cL377 . this change does not exist in scylladb/scylladb master. and i failed to find related explanation in the commit message of d0a60b1d6ffc5f73441fa2723ef89e478be18e2e . shall we port this change back to this repo? if yes, could you please provide some context of this change?

File: pgo/conf/clustering.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+#
+# Keyspace info
+#
+keyspace: keyspace1
+
+#
+# The CQL for creating a keyspace (optional if it already exists)
+#
+keyspace_definition: |
+  CREATE KEYSPACE keyspace1 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1};
```
Comment (mykaul) [inline pgo/conf/clustering.yaml:23]: unrelated to this PR - I wonder what code doesn't get LTO/PGO treatment by the fact we are using RF=1, which is not the common RF.

Comment (mykaul) [reply pgo/conf/clustering.yaml:23]: I see below we have other RFs.

File: cmake/enable_lto.cmake [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright 2024-present ScyllaDB
+#
+
+#
+# SPDX-License-Identifier: ScyllaDB-Proprietary
+#
```
Comment (raphaelsc) [inline cmake/enable_lto.cmake:7]: LicenseRef-ScyllaDB-Source-Available-1.0 ?

Comment (mykaul) [reply cmake/enable_lto.cmake:7]: I wonder if we can add a license check as a GH action.

Comment (raphaelsc) [reply cmake/enable_lto.cmake:7]: > I wonder if we can add a license check as a GH action.

That'd be great. /cc @yaronkaikov

Comment (tchaikov) [reply cmake/enable_lto.cmake:7]: > I wonder if we can add a license check as a GH action.

created at #22065

**Comment by tchaikov** [comment]
v4:

- replace two "ScyllaDB-Proprietary" SPDX IDs with the new "LicenseRef-ScyllaDB-Source-Available-1.0" ID.

File: cmake/enable_lto.cmake (hunk 2 above)
Comment (tchaikov) [reply cmake/enable_lto.cmake:7]: thank you Raphael for pointing this out. should be fixed in the latest revision.

**Review by michoecho — APPROVED** [review]
I did a shallow review, looks fine to me. If there are any problems, we will probably learn about them soon enough.

**Comment by tchaikov** [comment]
thank you Michał, for your review and approval.

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Comment by avikivity** [comment]
The subject is still wrong.

**Comment by tchaikov** [comment]
> The subject is still wrong.

@avikivity the subject of this PR should be better now. could you take another look?

**Comment by avikivity** [comment]
@yaronkaikov please set up the pgo refresh process once this promotes.

**Pull request merged**